[Merged by Bors] - feat: add `Filter.generate_image_preimage_le_comap`

pechersky opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Ruben-VandeVelde commented on Mathlib/Order/Filter/Finite.lean:
```diff
@@ -71,6 +71,15 @@ theorem mem_generate_iff {s : Set <| Set α} {U : Set α} :
     exact mem_of_superset ((sInter_mem tfin).2 fun V hV => GenerateSets.basic <| hts hV) h
 
+lemma generate_image_preimage_le_comap (U : Set (Set α)) (f : β → α) :
+  .generate ((f ⁻¹' ·) '' U) ≤ Filter.comap f (.generate U) := by
```
```suggestion
    .generate ((f ⁻¹' ·) '' U) ≤ Filter.comap f (.generate U) := by
```

ocfnash reviewed (approved):
Thanks!

Please apply something along the lines suggested in my code comment, and then feel free to merge.

bors d+

pechersky commented:
I moved the lemma, and added a similar what that seemed similar in
structure. Could you check that it matches what you were thinking?

On Thu, Mar 27, 2025 at 5:34 AM Oliver Nash ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> Thanks!
>
> Please apply something along the lines suggested in my code comment, and
> then feel free to merge.
>
> bors d+
> ------------------------------
>
> In Mathlib/Order/Filter/Finite.lean
> <https://github.com/leanprover-community/mathlib4/pull/23113#discussion_r2016065588>
> :
>
> > +lemma generate_image_preimage_le_comap (U : Set (Set α)) (f : β → α) :
> +    .generate ((f ⁻¹' ·) '' U) ≤ Filter.comap f (.generate U) := by
> +  rintro s ⟨t, ht⟩
> +  simp only [mem_generate_iff, exists_subset_image_iff, sInter_image] at ht ⊢
> +  obtain ⟨⟨u, hu, huf, hut⟩, hts⟩ := ht
> +  refine ⟨u, hu, huf.image _, subset_trans ?_ hts⟩
> +  rw [← preimage_sInter]
> +  exact preimage_mono hut
>
> I tend to think that having to manipulate individual elements in settings
> like this is usually a sign that API is missing or not being used. It also
> seems a shame that this result has to be in the Filter.Finite file.
>
> With these remarks in mind, how about placing this lemma in the Filter.Map
> file instead and using the following proof:
>
> lemma generate_image_preimage_le_comap (U : Set (Set α)) (f : β → α) :
>     generate ((f ⁻¹' ·) '' U) ≤ comap f (generate U) := by
>   rw [← map_le_iff_le_comap, le_generate_iff]
>   exact fun u hu ↦ mem_generate_of_mem ⟨u, hu, rfl⟩
>
> (This is perhaps overgolfed but it does make it clear that the lemma is
> really a consequence of three pieces of API, namely map_le_iff_le_comap,
> le_generate_iff, and mem_generate_of_mem.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/leanprover-community/mathlib4/pull/23113#pullrequestreview-2720512707>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIYHBSEXHPFJYQDHAWVGCL2WOZ2XAVCNFSM6AAAAABZLXYJN6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMRQGUYTENZQG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ocfnash commented on Mathlib/Order/Filter/Map.lean:
```diff
@@ -762,6 +762,16 @@ theorem inf_principal_eq_bot_iff_comap {F : Filter α} {s : Set α} :
   rw [principal_eq_map_coe_top s, ← Filter.push_pull', inf_top_eq, map_eq_bot_iff]
 
+lemma map_generate_le_generate_preimage_preimage (U : Set (Set β)) (f : β → α) :
+    map f (.generate U) ≤ .generate ((f ⁻¹' ·) ⁻¹' U) := by
```
These `.`s are redundant, right?
```suggestion
    map f (generate U) ≤ generate ((f ⁻¹' ·) ⁻¹' U) := by
```
Similarly below

pechersky commented:
bots r+
